### PR TITLE
feat: add fnn functionality

### LINKED ISSUE #522: Add FNN-Functionality
### Is your feature request related to a problem?

Create new branch because my test branch is impossible to merge

### Desired solution

Create new branch

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by sibre28**
Closes #522 

### Summary of Changes

Added Model and layer classes to support Feed Forward Neural Network Functionality.
Added into_dataloader() Function in TaggedTable Class, that creates a pytorch dataloader for any given tagged table, which can then be used to train a FNN.
Tests for those modules are probably incomplete, as i wasnt really sure what to test for, could use some feedback there.

### CODE REVIEW

File: src/safeds/ml/nn/_model.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                accuracies_test.append(accuracy.item())
+
+        # print(np.mean(loss_values_test))
+        # print(np.mean(accuracies_test))
+
+    def is_for_regression(self) -> bool:
+        return self._model.last_layer_has_output_size_one()
+
+
+class PytorchModel(nn.Module):
```
Comment (Marsmaennchen221) [inline src/safeds/ml/nn/_model.py:100]: This should only be an internal class, so consider using a leading underscore

File: src/safeds/ml/nn/_model.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        with torch.no_grad():
+            for x, y in dataloader:
+                pred = self._model(x)
+                loss = loss_fn(pred, y)
+                loss_values_test.append(loss.item())
+                accuracy = torch.mean(1 - torch.abs(pred - y))
+                accuracies_test.append(accuracy.item())
+
+
+class ClassificationModel:
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:90]: Use the [`Classifier`](https://library.safeds.com/en/stable/reference/safeds/ml/classical/classification/Classifier/#safeds.ml.classical.classification.Classifier) as the baseline for your interface.

File: src/safeds/ml/nn/_model.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                accuracy = torch.mean(1 - torch.abs(pred - y))
+                tmp_accuracies.append(accuracy.item())
+
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+            loss_values.append(np.mean(tmp_loss))
+            accuracies.append(np.mean(tmp_accuracies))
+
+    def predict(self, test_data: TaggedTable) -> None:
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:146]: Input must be a `Table`, it's not required to be a `TaggedTable`. The return value is a `TaggedTable`, not `None`.

File: src/safeds/ml/nn/_model.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        Parameters
+        ----------
+        test_data : TaggedTable
+            The data the network should try to predict.
+
+        """
+        dataloader = test_data.into_dataloader(self._batch_size)
+        self._model.eval()
+        loss_values_test = []
+        accuracies_test = []
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:159]: What is the purpose of computing loss values and accuracies in the `predict` method?

File: src/safeds/ml/nn/_model.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        loss_fn = nn.MSELoss()
+        with torch.no_grad():
+            for x, y in dataloader:
+                pred = self._model(x)
+                loss = loss_fn(pred, y)
+                loss_values_test.append(loss.item())
+                accuracy = torch.mean(1 - torch.abs(pred - y))
+                accuracies_test.append(accuracy.item())
+
+    def is_for_regression(self) -> bool:
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:169]: Seems out of place in a `ClassifierModel` class.

File: src/safeds/ml/nn/_model.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
+import numpy as np
+import torch
+from torch import nn
+
+from safeds.data.tabular.containers import TaggedTable
+from safeds.exceptions import ClosedBound, OutOfBoundsError
+from safeds.ml.nn._fnn_layer import FNNLayer
+
+
+class RegressionModel:
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:10]: Name should clarify, that this is a neural network, not just any model.

File: src/safeds/ml/nn/_model.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+                accuracy = torch.mean(1 - torch.abs(pred - y))
+                accuracies_test.append(accuracy.item())
+
+
+class ClassificationModel:
+    def __init__(self, layers: list):
+        self._model = _PytorchModel(layers, is_for_classification=True)
+        self._batch_size = 1
+
+    def train(self, train_data: TaggedTable, epoch_size: int = 25, batch_size: int = 1) -> None:
```
[the remark this answers is not in the document]
Comment (lars-reimann) [reply src/safeds/ml/nn/_model.py:95]: The function must return the trained model and not `None`. The original model must not be mutated.

File: src/safeds/ml/nn/_model.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+from safeds.exceptions import ClosedBound, OutOfBoundsError
+from safeds.ml.nn._fnn_layer import FNNLayer
+
+
+class RegressionModel:
+    def __init__(self, layers: list):
+        self._model = _PytorchModel(layers, is_for_classification=False)
+        self._batch_size = 1
+
+    def train(self, train_data: TaggedTable, epoch_size: int = 25, batch_size: int = 1) -> None:
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:15]: Try to be consistent with the other models. There, the function is called fit.

File: src/safeds/ml/nn/_model.py (hunk 6 above)
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:10]: Use the [Regressor](https://library.safeds.com/en/stable/reference/safeds/ml/classical/regression/Regressor/) abstract class as the baseline for your interface.

File: src/safeds/ml/nn/_model.py (hunk 8 above)
Comment (lars-reimann) [reply src/safeds/ml/nn/_model.py:15]: The function must return the trained model and not None. The original model must not be mutated.

File: src/safeds/ml/nn/_model.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        self._batch_size = batch_size
+        dataloader = train_data.into_dataloader(self._batch_size)
+
+        loss_fn = nn.BCELoss()
+
+        optimizer = torch.optim.SGD(self._model.parameters(), lr=0.05)
+
+        loss_values = []
+        accuracies = []
+        for _epoch in range(epoch_size):
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:128]: Underscore not needed for local variable.

File: src/safeds/ml/nn/_model.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+                accuracy = torch.mean(1 - torch.abs(pred - y))
+                tmp_accuracies.append(accuracy.item())
+
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+            loss_values.append(np.mean(tmp_loss))
+            accuracies.append(np.mean(tmp_accuracies))
+
+    def predict(self, test_data: TaggedTable) -> None:
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:66]: Input must be a Table, it's not required to be a TaggedTable. The return value is a TaggedTable, not None.

**Review by lars-reimann — CHANGES_REQUESTED** [review]

File: src/safeds/ml/nn/_fnn_layer.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+        """
+        if input_size < 1:
+            raise OutOfBoundsError(actual=input_size, name="input_size", lower_bound=ClosedBound(1))
+        if output_size < 1:
+            raise OutOfBoundsError(actual=output_size, name="output_size", lower_bound=ClosedBound(1))
+        self.input_size = input_size
+        self.output_size = output_size
+
+    def get_pytorch_layer(self, is_for_classification: bool) -> _PytorchLayer:
```
[the remark this answers is not in the document]
Comment (sibre28) [reply src/safeds/ml/nn/_fnn_layer.py:49]: Agreed, changed it to be a private method. The `is_for_classification` Flag is set in the init functions of ´RegressionModel` and `ClassificationModel` and is used to determine the activation function of a layer.

File: src/safeds/ml/nn/_model.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+from safeds.ml.nn._fnn_layer import FNNLayer
+
+
+class RegressionNeuralNetwork:
+    def __init__(self, layers: list):
+        self._model = _PytorchModel(layers, is_for_classification=False)
+        self._batch_size = 1
+        self._is_fitted = False
+
+    def fit(self, train_data: TaggedTable, epoch_size: int = 25, batch_size: int = 1) -> Self:
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:19]: * Add callback `on_epoch_completion` callback that gets called whenever an epoch is done.
* Add callback `on_batch_completion` callback that gets called whenever a batch is done.

File: src/safeds/ml/nn/_model.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+        return self._is_fitted
+
+
+class ClassificationNeuralNetwork:
+    def __init__(self, layers: list):
+        self._model = _PytorchModel(layers, is_for_classification=True)
+        self._batch_size = 1
+        self._is_fitted = False
+
+    def fit(self, train_data: TaggedTable, epoch_size: int = 25, batch_size: int = 1) -> Self:
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:130]: Callbacks for epoch & batch completion.

File: src/safeds/ml/nn/_model.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+
+        loss_fn = nn.MSELoss()
+
+        optimizer = torch.optim.SGD(copied_model._model.parameters(), lr=0.05)
+
+        loss_values = []
+        accuracies = []
+        for _ in range(epoch_size):
+            tmp_loss = []
+            tmp_accuracies = []
```
Comment (Marsmaennchen221) [inline src/safeds/ml/nn/_model.py:61]: Consider using a running loss/accuracy which is the current sum of the loss/accuracy in the epoch. You can get the average by dividing through the total number of batches*batch_size

File: src/safeds/ml/nn/_model.py [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+        copied_model = copy.deepcopy(self)
+        copied_model._batch_size = batch_size
+        dataloader = train_data.into_dataloader(copied_model._batch_size)
+
+        loss_fn = nn.MSELoss()
+
+        optimizer = torch.optim.SGD(copied_model._model.parameters(), lr=0.05)
+
+        loss_values = []
+        accuracies = []
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:58]: Accuracy does not make sense for a regression model. There's also no need to keep track of a single metric if the callback are available. -> Remove `accuracies` completley.

File: src/safeds/ml/nn/_model.py [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+        copied_model = copy.deepcopy(self)
+        copied_model._batch_size = batch_size
+        dataloader = train_data.into_dataloader(copied_model._batch_size)
+
+        loss_fn = nn.BCELoss()
+
+        optimizer = torch.optim.SGD(copied_model._model.parameters(), lr=0.05)
+
+        loss_values = []
+        accuracies = []
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:169]: There's also no need to keep track of a single metric if the callback are available. -> Remove `accuracies` completley.

File: src/safeds/ml/nn/_model.py [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+            for x, y in dataloader:
+                pred = copied_model._model(x)
+
+                loss = loss_fn(pred, y)
+                tmp_loss.append(loss.item())
+
+                accuracy = torch.mean(1 - torch.abs(pred - y))
+                tmp_accuracies.append(accuracy.item())
+
+                optimizer.zero_grad()
```
Comment (Marsmaennchen221) [inline src/safeds/ml/nn/_model.py:71]: Check whether this function is at the right position

File: src/safeds/ml/nn/_model.py [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+                loss.backward()
+                optimizer.step()
+            loss_values.append(np.mean(tmp_loss))
+            accuracies.append(np.mean(tmp_accuracies))
+        copied_model._is_fitted = True
+        return copied_model
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:77]: Set the `eval` mode here before returning.

File: src/safeds/ml/nn/_model.py [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+
+                accuracy = torch.mean(1 - torch.abs(pred - y))
+                tmp_accuracies.append(accuracy.item())
+
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+            loss_values.append(np.mean(tmp_loss))
+            accuracies.append(np.mean(tmp_accuracies))
+        copied_model._is_fitted = True
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:187]: Set `eval` mode here before returning.

File: src/safeds/ml/nn/_model.py [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+            The given test_data with an added "prediction" column at the end
+
+        Raises
+        ------
+        ModelNotFittedError
+            If the model has not been fitted yet
+        """
+        if not self._is_fitted:
+            raise ModelNotFittedError
+        copied_model = copy.deepcopy(self)
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:102]: Copy should not be needed for prediction.

File: src/safeds/ml/nn/_model.py [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+            The given test_data with an added "prediction" column at the end
+
+        Raises
+        ------
+        ModelNotFittedError
+            If the Model has not been fitted yet
+        """
+        if not self._is_fitted:
+            raise ModelNotFittedError
+        copied_model = copy.deepcopy(self)
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:213]: Copy should not be needed for prediction.

File: src/safeds/ml/nn/_model.py [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+        dataloader = test_data.into_dataloader(copied_model._batch_size)
+        copied_model._model.eval()
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:215]: `eval` mode should already be set after fitting.

File: src/safeds/ml/nn/_model.py [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+        Raises
+        ------
+        ModelNotFittedError
+            If the model has not been fitted yet
+        """
+        if not self._is_fitted:
+            raise ModelNotFittedError
+        copied_model = copy.deepcopy(self)
+        dataloader = test_data.into_dataloader(copied_model._batch_size)
+        copied_model._model.eval()
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:104]: `eval` mode should already be set after fitting.

**Review by lars-reimann — CHANGES_REQUESTED** [review]

File: src/safeds/data/tabular/containers/_table.py [hunk 24]
```diff
@@ -2380,3 +2382,41 @@ def __dataframe__(self, nan_as_null: bool = False, allow_copy: bool = True):  #
...
+
+    def into_dataloader(self, batch_size: int) -> DataLoader:
+        """
+        Return a Dataloader for the data stored in this table, used for training neural networks.
+
+        The original table is not modified.
+
+        Parameters
+        ----------
+        batch_size : int
```
Comment (lars-reimann) [inline src/safeds/data/tabular/containers/_table.py:2394]:
```suggestion
        batch_size
```

Docstrings should no longer contain types.

File: src/safeds/data/tabular/containers/_table.py [hunk 25]
```diff
@@ -2380,3 +2382,41 @@ def __dataframe__(self, nan_as_null: bool = False, allow_copy: bool = True):  #
...
+            The size of data batches that should be loaded at one time.
+
+        Returns
+        -------
+        result : DataLoader
```
Comment (lars-reimann) [inline src/safeds/data/tabular/containers/_table.py:2399]:
```suggestion
        result :
```

The colon is needed for results.

File: src/safeds/data/tabular/containers/_tagged_table.py [hunk 26]
```diff
@@ -844,3 +848,42 @@ def transform_column(self, name: str, transformer: Callable[[Row], Any]) -> Tagg
...
+
+    def into_dataloader(self, batch_size: int) -> DataLoader:
+        """
+        Return a Dataloader for the data stored in this table, used for training neural networks.
+
+        The original table is not modified.
+
+        Parameters
+        ----------
+        batch_size : int
```
Comment (lars-reimann) [inline src/safeds/data/tabular/containers/_tagged_table.py:860]:
```suggestion
        batch_size
```

File: src/safeds/data/tabular/containers/_tagged_table.py [hunk 27]
```diff
@@ -844,3 +848,42 @@ def transform_column(self, name: str, transformer: Callable[[Row], Any]) -> Tagg
...
+            The size of data batches that should be loaded at one time.
+
+        Returns
+        -------
+        result : DataLoader
```
Comment (lars-reimann) [inline src/safeds/data/tabular/containers/_tagged_table.py:865]:
```suggestion
        result :
```

File: src/safeds/ml/nn/_fnn_layer.py [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class FNNLayer:
+    def __init__(self, output_size: int, input_size: int | None = None):
+        """
+        Create a FNN Layer.
+
+        Parameters
+        ----------
+        input_size : int
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_fnn_layer.py:26]:
```suggestion
        input_size
```

File: src/safeds/ml/nn/_fnn_layer.py [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+            The number of neurons in the previous layer
+        output_size : int
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_fnn_layer.py:28]:
```suggestion
        output_size
```

File: src/safeds/ml/nn/_fnn_layer.py [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+        return _InternalLayer(self._input_size, self._output_size, is_last_layer_of_classification_model)
+
+    @property
+    def output_size(self) -> int:
+        """
+        Get the output_size of this layer.
+
+        Returns
+        -------
+        result: int
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_fnn_layer.py:56]:
```suggestion
        result :
```

File: src/safeds/ml/nn/_model.py [hunk 31]
```diff
@@ -0,0 +1,248 @@
...
+        train_data : TaggedTable
+            The data the network should be trained on.
+        epoch_size : int
+            The number of times the training cycle should be done.
+        batch_size : int
+            The size of data batches that should be loaded at one time.
+        callback_on_batch_completion: Callable[]
+            Function used to view metrics while training. Gets called after a batch is completed.
+        callback_on_epoch_completion: Callable[]
+            Function used to view metrics while training. Gets called after an epoch is completed.
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:43]:
```suggestion
        train_data
            The data the network should be trained on.
        epoch_size
            The number of times the training cycle should be done.
        batch_size
            The size of data batches that should be loaded at one time.
        callback_on_batch_completion
            Function used to view metrics while training. Gets called after a batch is completed.
        callback_on_epoch_completion
            Function used to view metrics while training. Gets called after an epoch is completed.
```

File: src/safeds/ml/nn/_model.py [hunk 32]
```diff
@@ -0,0 +1,248 @@
...
+
+        Raises
+        ------
+        ValueError
+            If epoch_size < 1
+            If batch_size < 1
+
+        Returns
+        -------
+        RegressionNeuralNetwork
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:53]:
```suggestion
        trained_model :
```

File: src/safeds/ml/nn/_model.py [hunk 33]
```diff
@@ -0,0 +1,248 @@
...
+
+    def predict(self, test_data: Table) -> TaggedTable:
+        """
+        Make a prediction for the given test data.
+
+        The original Model is not modified.
+
+        Parameters
+        ----------
+        test_data : Table
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:94]:
```suggestion
        test_data
```

File: src/safeds/ml/nn/_model.py [hunk 34]
```diff
@@ -0,0 +1,248 @@
...
+            The data the network should predict.
+
+        Returns
+        -------
+        TaggedTable
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:99]:
```suggestion
        prediction :
```

File: src/safeds/ml/nn/_model.py [hunk 35]
```diff
@@ -0,0 +1,248 @@
...
+        return test_data.add_column(Column("prediction", predictions)).tag_columns("prediction")
+
+    @property
+    def is_fitted(self) -> bool:
+        """
+        Check if the model is fitted.
+
+        Returns
+        -------
+        is_fitted : bool
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:123]:
```suggestion
        is_fitted
```

File: src/safeds/ml/nn/_model.py [hunk 36]
```diff
@@ -0,0 +1,248 @@
...
+        The original model is not modified.
+
+        Parameters
+        ----------
+        train_data : TaggedTable
+            The data the network should be trained on.
+        epoch_size : int
+            The number of times the training cycle should be done
+        batch_size : int
+            The size of data batches that should be loaded at one time.
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:148]:
```suggestion
        train_data
            The data the network should be trained on.
        epoch_size
            The number of times the training cycle should be done
        batch_size
            The size of data batches that should be loaded at one time.
```

File: src/safeds/ml/nn/_model.py [hunk 37]
```diff
@@ -0,0 +1,248 @@
...
+
+    def predict(self, test_data: Table) -> TaggedTable:
+        """
+        Make a prediction for the given test data.
+
+        The original Model is not modified.
+
+        Parameters
+        ----------
+        test_data : Table
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:194]:
```suggestion
        test_data
```

File: src/safeds/ml/nn/_model.py [hunk 38]
```diff
@@ -0,0 +1,248 @@
...
+            The data the network should predict.
+
+        Returns
+        -------
+        TaggedTable
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:199]:
```suggestion
        prediction :
```

File: src/safeds/ml/nn/_model.py [hunk 39]
```diff
@@ -0,0 +1,248 @@
...
+        return test_data.add_column(Column("prediction", predictions)).tag_columns("prediction")
+
+    @property
+    def is_fitted(self) -> bool:
+        """
+        Check if the model is fitted.
+
+        Returns
+        -------
+        is_fitted : bool
```
Comment (lars-reimann) [inline src/safeds/ml/nn/_model.py:223]:
```suggestion
        is_fitted :
```

**Review by lars-reimann — CHANGES_REQUESTED** [review]

File: src/safeds/data/tabular/containers/_table.py [hunk 40]
```diff
@@ -2380,3 +2382,41 @@ def __dataframe__(self, nan_as_null: bool = False, allow_copy: bool = True):  #
         data_copy = self._data.reset_index(drop=True)
         data_copy.columns = self.column_names
         return data_copy.__dataframe__(nan_as_null, allow_copy)
+
+    def into_dataloader(self, batch_size: int) -> DataLoader:
```
Comment (Gerhardsa0) [inline src/safeds/data/tabular/containers/_table.py:2386]: this should be a private function, as the user will not use this function and work directly on Torch objects.

```_into_dataloader``` or mby add something like a neural network utils

**Review by lars-reimann — APPROVED** [review]
We might need to make some adjustments later, but for now this is good to go.

**Pull request merged**